model/PermissionsService.php

```diff
@@ -178,6 +346,55 @@ private function deduplicateActions(array $actions): array
         return $actions;
     }
 
+    /**
+     * @fixme Maybe should not be public, and/or we should move this responsibility to the task
+     *        and/or we should make the threshold configurable
+     */
+    public function isHugeClass(core_kernel_classes_Resource $resource, int $threshold): bool
+    {
+        if ($resource->isClass()) {
+            // Ensure we have a core_kernel_classes_Class instance
+            $class = $resource->getClass($resource->getUri());
+
+            return ($class->countInstances([], ['recursive' => true]) >= $threshold);
+        }
+
+        return false;
+    }
+
+    private function debug(string $format, ...$va_args): void
+    {
+        static $logger = null;
+        if (!$logger) {
+            $logger = $this->getLogger();
+        }
+
+        $logger->info(self::class . ': ' . vsprintf($format, $va_args));
+    }
+
+    public function getNestedResources(core_kernel_classes_Resource $resource): Generator
+    {
+        $queue = new SplQueue();
```

Comment:
FIXME: Use a stack instead, as that will use less memory (i.e. depth first instead of breath first search)